src/cpu/aarch64/acl_binary.hpp

```diff
@@ -0,0 +1,266 @@
+                min_op->configure(&acl_obj_->src0_tensor,
+                        &acl_obj_->src1_tensor, &acl_obj_->dst_tensor);
+                acl_obj_->binary_op = std::move(min_op);
+                break;
+            }
+            case alg_kind::binary_max: {
+                auto max_op = std::make_unique<arm_compute::NEElementwiseMax>();
+                max_op->configure(&acl_obj_->src0_tensor,
+                        &acl_obj_->src1_tensor, &acl_obj_->dst_tensor);
+                acl_obj_->binary_op = std::move(max_op);
+                break;
+            }
+            default: return status::runtime_error;
+        }
+
+        return status::success;
+    }
+
+    acl_binary_obj_t &get_acl_obj() const { return *acl_obj_; }
+
+    DNNL_DISALLOW_COPY_AND_ASSIGN(acl_binary_resource_t);
+
+private:
+    std::unique_ptr<acl_binary_obj_t> acl_obj_;
+}; // acl_binary_resource_t
+
+struct acl_binary_t : public primitive_t {
+    struct pd_t : public cpu_binary_pd_t {
+        using cpu_binary_pd_t::cpu_binary_pd_t;
+
+        DECLARE_COMMON_PD_T("acl", acl_binary_t);
+
+        pd_t(const binary_desc_t *adesc, const primitive_attr_t *attr,
+                const binary_pd_t *hint_fwd_pd)
+            : cpu_binary_pd_t(adesc, attr, hint_fwd_pd) {}
+
+        status_t init(engine_t *engine) {
+
+            using namespace acl_common_utils;
+
+            // Only support f32 and s32 for now
+            data_type_t ddt = dst_md(0)->data_type;
+            if (ddt != data_type::f32 && ddt != data_type::s32)
+                return status::unimplemented;
+
+            // Only support src and dst all matching for now
+            if (ddt != src_md(0)->data_type || src_md(1)->data_type != ddt)
+                return status::unimplemented;
+
+            // Sets the memory format of dst from any to src_md(0) blocking desc
+            CHECK(set_default_params());
+
+            if (!attr()->has_default_values()) return status::unimplemented;
+
+            asp_.alg = desc()->alg_kind;
+
+            // All the algorithms we support
+            if (!utils::one_of(asp_.alg, alg_kind::binary_add,
+                        alg_kind::binary_sub, alg_kind::binary_mul,
+                        alg_kind::binary_div, alg_kind::binary_max,
+                        alg_kind::binary_min))
+                return status::unimplemented;
+
+            // s32 div in ACL does not round as oneDNN expects
+            if (ddt == data_type::s32 && asp_.alg == alg_kind::binary_div)
+                return status::unimplemented;
+
+            // ACL pointwise arithmetic operators assume that the innermost
+            // dimensions are dense for src0, src1 and dst. So we permute a
+            // common dense dim to be the innermost one
+            memory_desc_t src_d0_permed, src_d1_permed, dst_d_permed;
+            CHECK(permute_common_dense_dimension_to_last(&src_d0_permed,
```

Comment:
QQ: could you, please, explain how this magic works? Based on function comment, let's consider the example you posted: "As an example for this function, if the input formats were {abdc, acdb, acdb} then the permuted dimensions would be {abcd, acbd, acbd}.". So if user provided `abdc` as src0, this function will transform it into `abcd`. User passes memory in `abdc` layout, what ACL library does in this case? I haven't noticed any reorders or logic that will actually convert one layout into another (or a spot which explains that this conversion is not needed and ACL magically just works). This part I'm missing. Could you shed a light on it, please? Thank you.

Reply:
The function permutes the logical dimensions using `dnnl_memory_desc_permute_axes`, so no reorder of the underlying data is required. It does mean that the _meaning_ of the dimensions changes, so for example if the channels dimension had an index of 2, then it may now have an index of 1. This is fine for binary ops, because we just need the logical dimensions of `src0`, `src1` and `dst` to match up.

I have changed the wording of this comment, hopefully it is clearer now.